api/dimensions.go

```diff
@@ -184,18 +128,13 @@ func convertBSONToDimensionOption(data interface{}) (*models.DimensionOption, er
 	return &dim, nil
 }
 
-func (api *DatasetAPI) getDimensionOptions(w http.ResponseWriter, r *http.Request) {
+func (api *DatasetAPI) getDimensionOptions(w http.ResponseWriter, r *http.Request, limit, offset int) (interface{}, int, error) {
```

Comment:
Question: Is there is a reason this method returns an interface type rather than a specific type? I am guessing this is allow the paginator to work?

This seems OK however I am slightly wary because this means these methods are no longer fulfilling the `http.Handler` interface which might cause problems further down the road. 

For example if we need to wrap this in another `http.Handler` which is common pattern to add functionality to an API we wouldn;t be able to do that in the same way any more. Or if we need to remove the pagaination layer it wouldn't be as simple as removing it from the API set up - this method isn't a handler anymore so there would be other changes necessary.


This may not be an issue and it may also be unavoidable however I think it should be discussed and made as a conscious decision because there is a potential impact and downside to doing this. Is there another way we can achieve the same outcome by defining an interface for a "paginating handler"?

Reply:
heya, with regards to the question, yeah - we want the paginator to work with any array of structs passed to it, we tried to use []interface as our return type however we couldn't get that to work.

Looking at the handler interface, yeah we have had to diverge here - we don't appear to have the ability with mux to work with a request scoped context so this is why we've had to move away from the pattern. 

We definitely want a separation of concerns between the logic of returning the correct items from the database and the rendering of a paginated HTTP response, which is why we've chosen to structure it in this way. Trouble is, limit and offset are very much parameters relevant to pagination, however we need to use them in the db layer to retrieve the correct items. If we had a request context we could keep this as a standard middleware layer quite easily and use that to communicate those values.

I would suggest that if in the future more middleware for these endpoints was required, that moving to something like Gin or Echo is considered which would make this kind of thing straightforward